core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1966,29 +1966,79 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      fibers: Iterable[Fiber[E, _]],
+      promise: Promise[E, List[B]],
+      buffer: Array[B],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.halt(e).unit *> Fiber.interruptAll(fibers),
+        a =>
+          todo.modify { t =>
+            buffer.update(idx, a)
+            if (t == 1) promise.succeed(buffer.toList).unit -> 0 else UIO.unit -> (t - 1)
+          }.flatten
+      )
+
+    (for {
+      size    <- UIO.effectTotal(as.size)
+      todo    <- Ref.make(size)
+      buffer  <- UIO.effectTotal(new Array[Any](size).asInstanceOf[Array[B]])
+      promise <- Promise.make[E, List[B]]
+      c <- ZIO.uninterruptibleMask { restore =>
+            for {
+              as <- ZIO.traverse(as)(a => ZIO.interruptible(fn(a)).fork)
+              _ <- as.view.zipWithIndex.foldLeft[ZIO[R, E, _]](ZIO.unit) {
+                    case (io, (f, idx)) =>
+                      io *> f.await.flatMap(arbiter(as, promise, buffer, todo, idx)).fork
+                  }
+              _ <- promise.succeed(Nil).when(as.isEmpty)
+              c <- restore(promise.await).onInterrupt(promise.interrupt *> Fiber.interruptAll(as))
+            } yield c
+          }
+    } yield c).refailWithTrace
+  }
 
   /**
    * Applies the function `f` to each element of the `Iterable[A]` and runs
    * produced effects in parallel, discarding the results.
    *
    * For a sequential version of this method, see `foreach_`.
    */
-  final def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, _]): ZIO[R, E, Unit] =
-    ZIO
-      .succeed(as.iterator)
-      .flatMap { i =>
-        def loop(a: A): ZIO[R, E, Unit] =
-          if (i.hasNext) f(a).zipWithPar(loop(i.next))((_, _) => ())
-          else f(a).unit
-        if (i.hasNext) loop(i.next)
-        else ZIO.unit
-      }
-      .refailWithTrace
+  final def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, _]): ZIO[R, E, Unit] = {
```

Comment:
This function is duplicating a lot from the foreachPar method, maybe this could be rewritten to leverage that function instead?

Reply:
I know it's possible, I just wanted to avoid allocation of the extra array with results.